merge-queue blocks "unreviewed" PRs if r+ received from a user that isn't a reviewer

briannafan opened the pull request:
#### d13948d3347b05562a92cc1a91235f3767c85af8
<pre>
merge-queue blocks &quot;unreviewed&quot; PRs if r+ received from a user that isn&apos;t a reviewer
<a href="https://bugs.webkit.org/show_bug.cgi?id=277038">https://bugs.webkit.org/show_bug.cgi?id=277038</a>
<a href="https://rdar.apple.com/132435062">rdar://132435062</a>

Reviewed by Aakash Jain.

Ignore invalid reviewers in ValidateCommitterAndReviewer. They will be caught in ValidateCommitMessage.
This allows for invalid reviews on commits that don&apos;t need reviews  and if a valid reviewer is
only written in the commit message.

* Tools/CISupport/ews-build/factories_unittest.py:
(TestExpectedBuildSteps): Fix typos.

* Tools/CISupport/ews-build/steps.py:
(ValidateChange.validate_github): Fix typo.
(ValidateCommitterAndReviewer): Ignore invalid reviewer. Set &apos;invalid_reviewers&apos; and &apos;valid_reviewers&apos;.
(ValidateCommitterAndReviewer.run):
(ValidateCommitMessage.run): Make summary more detailed.
(AddReviewerMixin.reviewers): Replace &apos;reviewers_full_names&apos; with &apos;valid_reviewers&apos;.
(AddReviewerToCommitMessage.doStepIf): Ditto.

* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/282024@main">https://commits.webkit.org/282024@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d0e2a2cbcfad7563e71dfba5544e03a2940aba10

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61795 "6 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41149 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14387 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/65774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/12340 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63914 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48835 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12611 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/65774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/12340 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64864 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38215 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53522 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/65774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34866 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10753 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/11271 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56674 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11056 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/67503 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5738 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/10802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/67503 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/61473 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5763 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53470 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/67503 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4712 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9310 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/36949 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/38033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/39129 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/37778 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

bls1999 commented on Tools/CISupport/ews-build/factories_unittest.py:
```diff
@@ -701,7 +701,7 @@ class TestExpectedBuildSteps(unittest.TestCase):
             'configure-build',
             'validate-change',
-            'validate-commiter-and-reviewer',
+            'validate-committer-and-reviewer',
```
Change of the year! 🎉

bls1999 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -1929,7 +1957,7 @@ def validate_github(self, pr_number):
 
         if pr_json:
-            # Only track acionable labels, since bug category labels may reveal information about security bugs
+            # Only track actionable labels, since bug category labels may reveal information about security bugs
```
Another awesome change.

bls1999 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -1971,6 +1999,14 @@ def validate_github(self, pr_number):
+        commit_data = yield self.get_pr_commits(pr_number, repository_url, retry=3)
+        commit_msg = commit_data[0].get('commit', {}).get('message', '')
+        for line in commit_msg.lower().split():
+            if line.startswith('unreviewed'):
```
I'm wondering if the way the commit message is currently checked is a better idea here?

briannafan replied:
Could you please elaborate on "the way the commit message is currently checked"?
We grep for 'Unreviewed' in `ValidateCommitMessage` which is essentially equivalent to this method.

bls1999 replied:
Yes, happy to clarify. My comment was about the use of `startswith` specifically. Consider the case of some accidental whitespace at the beginning of a line, for instance. It's probably not hugely important, just a thought I had since we were discussing the grep.

briannafan replied:
To follow the spirit of `ValidateCommitMessage`, I'll change it to searching the entire message for 'Unreviewed' (case sensitive).

bls1999 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2155,11 +2191,17 @@ def filter_out_non_reviewer_validators(candidate):
-                defer.returnValue(rc)
-                return
+                if unreviewed:
+                    yield self._addToLog('stdio', f'PR is marked as unreviewed. {reviewer} is not a valid reviewer.\n')
```
Nit, but what about: `{reviewer} is not a valid reviewer, but the PR is marked as unreviewed.\n`

This makes it a bit clearer, I think.

aj062 reviewed (changes requested):
This approach works, but is very limited. It doesn't cover many cases which ValidateCommitMessage step covers, 'Unreviewed' is just one of those cases. There are other cases like: 
- Rubber-stamped
- Versioning
- when the commit message has the string “Reviewed by xyz”, but that reviewer didn’t add the review in Github (that’s a valid review method as there might be scenarios like the person did over the shoulder review, or had issue with GitHub account etc.)


I would like to propose a much simpler approach which cover all these cases: “Simply ignore the review from a non-reviewer”. The implementation is also much simpler. 

This is how it works:
- While going through the list of reviewers on a PR, if we encounter a person without reviewer status, instead of failing the build, simply ignore that reviewer (and add a log).
- At the end, if there is no valid reviewer found, ValidateCommitterAndReviewer step would succeed just like it does when there is no reviewer on the PR (e.g: Unreviewed PR). The subsequent step of ValidateCommitMessage would check the commit message and will fail the build since no valid reviewer was found.

Here is sample patch for that logic (note that this is untested).

[ignore_invalid_reviewer.patch](https://github.com/user-attachments/files/16446645/ignore_invalid_reviewer.patch)

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6569,8 +6572,12 @@ def _files(self):
     def run(self, BufferLogObserverClass=logobserver.BufferLogObserver):
         base_ref = self.getProperty('github.base.ref', f'{DEFAULT_REMOTE}/{DEFAULT_BRANCH}')
         head_ref = self.getProperty('github.head.ref', 'HEAD')
-        reviewers = self.getProperty('reviewers_full_names', None)
```
`reviewers_full_names` property is renamed in this PR, but all the places which use this property doesn't seem to be updated. This would break Merge-Queue (e.g.: AddReviewerToCommitMessage would stop adding reviewer name to commit message and then ValidateCommitMessage will fail).

briannafan replied:
Great catch! I meant to do that but forgot.. I'm also going to be testing the changes on DEV today.

briannafan replied:
Here's an example of AddReviewerToCommitMessage working as intended: https://ews-build.webkit-dev.org/#/builders/38/builds/35825

briannafan replied:
Example of no valid reviewer: https://ews-build.webkit-dev.org/#/builders/38/builds/35826

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2155,25 +2155,28 @@ def filter_out_non_reviewer_validators(candidate):
-        if any([self.is_reviewer(reviewer) for reviewer in reviewers]):
-            reviewers = list(filter(self.is_reviewer, reviewers))
-        reviewers = set(reviewers)
+        valid_reviewers = []
```
Actually, can make the code little compact by initializing it as set() instead of list (and using .add() instead of .append()).

aj062 commented on Tools/CISupport/ews-build/steps_unittest.py:
```diff
@@ -8096,6 +8102,30 @@ def test_success_period(self):
+            ExpectShell(workdir='wkdir',
+                        logEnviron=False,
+                        timeout=60,
+                        command=['/bin/sh', '-c', "git log eng/pull-request-branch ^main | grep -q 'OO*PP*S!' && echo 'Commit message contains (OOPS!) and Web Kit is not a reviewer' || test $? -eq 1"])
```
Do we need to add this: '-o', 'pipefail'

aj062 reviewed (approved):
Looks good.
Please update unit-tests.

briannafan committed:
merge-queue blocks "unreviewed" PRs if r+ received from a user that isn't a reviewer
https://bugs.webkit.org/show_bug.cgi?id=277038
rdar://132435062

Reviewed by Aakash Jain.

Ignore invalid reviewers in ValidateCommitterAndReviewer. They will be caught in ValidateCommitMessage.
This allows for invalid reviews on commits that don't need reviews  and if a valid reviewer is
only written in the commit message.

* Tools/CISupport/ews-build/factories_unittest.py:
(TestExpectedBuildSteps): Fix typos.

* Tools/CISupport/ews-build/steps.py:
(ValidateChange.validate_github): Fix typo.
(ValidateCommitterAndReviewer): Ignore invalid reviewer. Set 'invalid_reviewers' and 'valid_reviewers'.
(ValidateCommitterAndReviewer.run):
(ValidateCommitMessage.run): Make summary more detailed.
(AddReviewerMixin.reviewers): Replace 'reviewers_full_names' with 'valid_reviewers'.
(AddReviewerToCommitMessage.doStepIf): Ditto.

* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: https://commits.webkit.org/282024@main
